src/car_backed_blockstore.rs

```diff
@@ -0,0 +1,272 @@
+///
+/// See [module documentation](mod@self) for more.
+#[derive(Debug)]
+struct BlockLocation {
+    offset: u64,
+    length: usize,
+}
+
+/// See [module documentation](mod@self) for more.
+// Theoretically, this should be clonable, with very low overhead
+pub struct CarBackedBlockstore<ReaderT> {
+    // Blockstore methods take `&self`, so lock here
+    reader: Mutex<ReaderT>,
+    write_cache: Mutex<AHashMap<Cid, Vec<u8>>>,
+    index: AHashMap<Cid, BlockLocation>,
+    pub roots: Vec<Cid>,
+}
+
+impl<ReaderT> CarBackedBlockstore<ReaderT>
+where
+    ReaderT: BufRead + Seek,
+{
+    /// To be correct:
+    /// - `reader` must read immutable data. e.g if it is a file, it should be [`flock`](https://linux.die.net/man/2/flock)ed.
+    ///   [`Blockstore`] API calls may panic if this is not upheld.
+    /// - `reader`'s buffer should have enough room for the [`CarHeader`] and any [`Cid`]s.
+    #[tracing::instrument(skip_all)]
+    pub fn new(mut reader: ReaderT) -> cid::Result<Self> {
+        let CarHeader { roots, version } = read_header(&mut reader)?;
+        if version != 1 {
+            return Err(cid::Error::Io(io::Error::new(
+                Unsupported,
+                "file must be CARv1",
+            )));
+        }
+
+        // now create the index
+        let index = std::iter::from_fn(|| read_block(&mut reader))
+            .collect::<Result<AHashMap<_, _>, _>>()?;
+        match index.len() {
+            0 => {
+                return Err(cid::Error::Io(io::Error::new(
+                    InvalidData,
+                    "CARv1 files must not be empty",
+                )))
+            }
+            num_blocks => debug!(num_blocks, "indexed CAR"),
+        }
+
+        Ok(Self {
+            reader: Mutex::new(reader),
+            index,
+            roots,
+            write_cache: Mutex::new(AHashMap::new()),
+        })
+    }
+}
+
+impl<ReaderT> Blockstore for CarBackedBlockstore<ReaderT>
+where
+    ReaderT: Read + Seek,
+{
+    // This function should probably return a Cow<[u8]> to save unneccessary memcpys
+    #[tracing::instrument(skip(self))]
+    fn get(&self, k: &Cid) -> anyhow::Result<Option<Vec<u8>>> {
+        match (self.index.get(k), self.write_cache.lock().entry(*k)) {
+            (Some(_location), Occupied(cached)) => {
+                debug!("evicting from write cache");
+                Ok(Some(cached.remove()))
+            }
+            (Some(BlockLocation { offset, length }), Vacant(_)) => {
+                debug!("fetching from disk");
+                let mut reader = self.reader.lock();
+                reader.seek(SeekFrom::Start(*offset))?;
+                let cid = Cid::read_bytes(&mut *reader)?;
+                assert_eq!(cid, *k);
+                let cid_len = reader.stream_position()? - *offset;
+                let data_len = *length - usize::try_from(cid_len).unwrap();
+                let mut data = vec![0; data_len];
```

Comment:
Why do we need to initialize it with zeroes? Given that this chunk of memory will get populated in the next line, this may be redundant.

Reply:
> Why do we need to initialize it with zeroes

In short, because rust is a memory-safe language.

You're not allowed to observe uninitialised memory (though the compiler is allowed to optimise memory that you never observe away).

We could jump through hoops, with 
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len or https://doc.rust-lang.org/std/mem/union.MaybeUninit.html

But I don't know if it's worth it yet

Reply:
Looks good as is to me, some context: 
https://github.com/rust-lang/rust/pull/40409
https://users.rust-lang.org/t/is-set-len-the-way-to-make-an-unintialized-vec-as-a-buffer/38248/4

Reply:
I'm not convinced. `calloc` is more expensive than `malloc`, right? I mean, if there is no clean way to do it nicely in Rust then I'm good with this, but requesting zero-initialized memory is wasteful from what I see here.

Reply:
There is no alternative in safe rust :shrug: